### PR TITLE
Add a WithSecrets to App installer

### LINKED ISSUE #267: Add WithSecrets to app installer
It would be useful to create a func to install secrets in the new app installer

Something like .WithSecrets() that creates the secret as part of the install commands.

### PR DESCRIPTION
**Pull request by Waterdrips**
## Description

This converts OpenFaaS to the new app style and adds a .WithSecret() to
app options. This can be used to create a basic secret for now (the
default k8s type of generic) - It could be extended in the future to
create other k8s secret types.

Signed-off-by: Alistair Hey <alistair@heyal.co.uk>

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Fixes #267 

## How Has This Been Tested?

Tested by installing OpenFaaS On a k3d cluster and logging in to the UI
(thich uses the basic auth secret)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

- [ ] I have tested this on arm, or have added code to prevent deployment

### CODE REVIEW

File: cmd/apps/openfaas_app.go [hunk 1]
```diff
@@ -122,23 +124,13 @@ func MakeInstallOpenFaaS() *cobra.Command {
...
 			}
-
-			res, secretErr := k8s.KubectlTask("-n", namespace, "create", "secret", "generic",
-				"basic-auth",
-				"--from-literal=basic-auth-user=admin",
-				`--from-literal=basic-auth-password=`+pass)
-
-			if secretErr != nil {
-				return secretErr
+			secretData := map[string]string{
```
Comment (alexellis) [inline cmd/apps/openfaas_app.go:127]: Maybe inline it?

File: pkg/apps/chart_app.go [hunk 2]
```diff
@@ -13,6 +14,12 @@ func MakeInstallChart(options *types.InstallerOptions) (*types.InstallerOutput,
 		return nil, err
 	}
 
+	for _, secret := range options.Secrets {
```
Comment (alexellis) [inline pkg/apps/chart_app.go:17]: Do you want to exit? We used to warn when the secret existed

File: pkg/k8s/kubernetes_exec.go [hunk 3]
```diff
@@ -50,3 +52,31 @@ func Kubectl(parts ...string) error {
...
+	args := []string{"-n", secret.Namespace, "create", "secret", secret.Type, secret.Name}
+	args = append(args, secretData...)
+
+	res, secretErr := KubectlTask(args...)
+
+	if secretErr != nil {
+		return secretErr
+	}
+	if res.ExitCode != 0 {
+		fmt.Printf("[Warning] unable to create secret %s, may already exist: %s", "basic-auth", res.Stderr)
```
Comment (alexellis) [inline pkg/k8s/kubernetes_exec.go:68]: Should this have `basic-auth"` hard-coded?

File: cmd/apps/openfaas_app.go [hunk 4]
```diff
@@ -54,6 +57,8 @@ func MakeInstallOpenFaaS() *cobra.Command {
 	openfaas.Flags().StringArray("set", []string{}, "Use custom flags or override existing flags \n(example --set=gateway.replicas=2)")
 
 	openfaas.RunE = func(command *cobra.Command, args []string) error {
+		openFaaSOptions := types.DefaultInstallOptions()
```
Comment (alexellis) [inline cmd/apps/openfaas_app.go:60]: For brevity `appOps`?

**Review by alexellis — COMMENTED** [review]
Sending a few comments / questions.

File: pkg/k8s/kubernetes_exec.go (hunk 3 above)
Comment (Waterdrips) [reply pkg/k8s/kubernetes_exec.go:71]: oops :eyes:

File: pkg/apps/chart_app.go (hunk 2 above)
Comment (Waterdrips) [reply pkg/apps/chart_app.go:17]: This bubbles up the `secretErr` that we used to error on, when the command has an exit code != 0 we just print a log line
```go 
	res, secretErr := KubectlTask(args...)

	if secretErr != nil {
		return secretErr
	}
```

File: cmd/apps/openfaas_app.go [hunk 5]
```diff
@@ -122,23 +124,13 @@ func MakeInstallOpenFaaS() *cobra.Command {
...
 			}
-
-			res, secretErr := k8s.KubectlTask("-n", namespace, "create", "secret", "generic",
-				"basic-auth",
-				"--from-literal=basic-auth-user=admin",
-				`--from-literal=basic-auth-password=`+pass)
-
-			if secretErr != nil {
-				return secretErr
+			secretData := []types.SecretsData{
```
Comment (alexellis) [inline cmd/apps/openfaas_app.go:127]: Wonder if you could use a fluent API for this too?

File: cmd/apps/openfaas_app.go [hunk 6]
```diff
@@ -122,23 +124,13 @@ func MakeInstallOpenFaaS() *cobra.Command {
...
 			}
-			if res.ExitCode != 0 {
-				fmt.Printf("[Warning] unable to create secret %s, may already exist: %s", "basic-auth", res.Stderr)
-			}
-		}
 
-		err = helm.FetchChart("openfaas/openfaas", defaultVersion)
-		if err != nil {
-			return err
+			basicAuthSecret := types.NewGenericSecret("basic-auth", namespace, secretData)
```
Comment (alexellis) [inline cmd/apps/openfaas_app.go:132]: Is there a way I can re-use this in the Venafi apps in #266 without using helm?

**Review by alexellis — APPROVED** [review]
Approved

**Pull request merged**